### ISSUE TITLE
how to reduce GPU memory

### ISSUE THREAD

**Issue by CLoudSnowMan** [Xiangyu-CAS/AICity2020-VOC-ReID#7]
GeForce GTX 1080
When I set FP16=True or false, the GPU memory doesn't change

**Comment by Xiangyu-CAS** [comment]
Hi, tensor core is needed for apex and mix precision training, but unfortunitly GTX series don't support it.  I have also runned this project on 1080Ti, the memory can be reduced at most 30% but observing no accelerate.

If FP16 mode is successfully turned off, the following information would be outputed. 

```
Defaults for this optimization level are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic
```


BTW, here are some tips for saving GPU memory:
1. set MODEL.LAST_STRIDE = 1, with accuracy loss of course.
2. use small but competitive backbone like regnet, RegNetY800M achieves similar performance to ResNet50 but is inferior to ResNet50-ibn-a, [pretrain model](https://github.com/facebookresearch/pycls)
```
MODEL.NAME "('regnety_800mf')" \
MODEL.PRETRAIN_PATH "('/home/zxy/.cache/torch/checkpoints/RegNetY-800MF_dds_8gpu.pyth')" 
```

**Comment by CLoudSnowMan** [comment]
Thanks for your reply! Now I have 50K IDs for training. I set NUM_INSTANCE=4 IMS,_PER_BATCH=32, After "open all layers", It throwed "cuda out of memorry".
First how can I use multi gpus instead of one gpu?
Second I found computing circle loss  cost 600M memorry, Is there any way to move to cpu to compute?
Third fp16 only uses in these places:
  if cfg.SOLVER.FP16:
            with amp.scale_loss(loss, optimizer) as scaled_loss:
                scaled_loss.backward()
If we can use fp16 in other place such as input data or model inference?

**Comment by Xiangyu-CAS** [comment]
Well, the first epoch turns off the backward of feature layers so it costs less memory
1. multi-gpu suffers from accuracy loss beacuse of sync-BN, so I just leave it in my TODO lists for a very long time. I would like to implement it in the furture but currently it's unsupported.

2. Loss computation seems not possible to be moved into CPU because auto-grad relies on it.

3. Actually, the whole inference and backward work in FP16 mode after setting "scaled_loss.backward()". If you choose a full FP16 mode you can observe the memory be cut off in half, but here we set it as mix-precision mode, like convolution in FP16 and BN in FP32.

**Comment by CLoudSnowMan** [comment]
Thanks for your reply!
First I understand the keypoint is Cross-GPU Synchronized BN!
Second in your code I found 
     score, feat = model(input1, target)
     loss = loss_fn(score, feat, target)
So I argue that the whole inference work in normal mode.
But the backward work in FP16 mode?
Third what about convolution in FP16 and BN in FP32?

**Comment by Xiangyu-CAS** [comment]
Haha, no worries, I am pretty sure the whole inference works in FP16 mode. At previous version of apex, model needs to be converted like "model.half()" and "data.half()", but now for convenient, it is done implicitly.  You can debug it and print the types. 

Convolution in FP16 and BN in FP32 means it is mix-precision mode, actually pure FP16 mode leads to accuracy loss ! (but setting opt_level=O3), you can check [offical docs](https://nvidia.github.io/apex/amp.html) for details

**Comment by CLoudSnowMan** [comment]
Thanks for your reply. Now I look forward to the multi-GPU  version. 
Refer to https://github.com/rwightman/pytorch-image-models/, it can do 'Training in distributed mode with multiple processes'

**Comment by Xiangyu-CAS** [comment]
Thank you for the reference. That's really a awesome project : ) . If you are using GTX series with apex disabled and cant wait to get a mult-gpu version, it's a good choice to use [fast-reid](https://github.com/JDAI-CV/fast-reid). Xingyu did really great work in this project, sync-BN was carefully handled without any accuracy loss.

**Comment by CLoudSnowMan** [comment]
Thanks for your opinion!
1、If I want to train with multi-gpus and not distributed, just add this code when I refer to fast-reid and is it OK? 
    # For training, wrap with DP. But don't need this for inference.
        model = DataParallel(model)
        if cfg.MODEL.BACKBONE.NORM == "syncBN":
            # Monkey-patching with syncBN
            patch_replication_callback(model)
        model = model.cuda()
2、Is there something else I should watch out?
